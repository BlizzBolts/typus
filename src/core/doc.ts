import ts from "typescript";
import type { SourceFile } from "./sourceFile";

export interface SerialziedDoc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: SerialziedDoc[];
  parameters?: SerialziedDoc[];
  children?: SerialziedDoc[];
  symbol?: ts.Symbol;
}

export interface DocPayload {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;
}

export class Doc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;

  root?: SourceFile | null;

  constructor(param: DocPayload) {
    this.name = param.name || "";
    this.documentation = param.documentation || "";
    this.type = param.type || "";
    this.tags = param.tags || [];
    this.required = param.required || false;
    this.members = param.members || [];
    this.parameters = param.parameters || [];
    this.children = param.children || [];
    this.symbol = param.symbol || (null as unknown as ts.Symbol);

    this.root = null;
  }

  public setRoot(root: SourceFile) {
    this.root = root;
  }

  static serialize = (curr: Doc): SerialziedDoc => {
    return {
      name: curr.name,
      documentation: curr.documentation,
      type: curr.type,
      tags: curr.tags,
      required: curr.required,
      members: curr.members?.map(Doc.serialize),
      parameters: curr.parameters?.map(Doc.serialize),
      children: curr.children?.map(Doc.serialize),
    };
  };
}
