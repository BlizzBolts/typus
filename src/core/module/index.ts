export { ModuleGraph } from "./graph";
export { Module } from "./module";
export type { BaseModule } from "./types";
