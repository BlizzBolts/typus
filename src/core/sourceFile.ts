import { Doc, SerialziedDoc } from "./doc";

export type FilePath = string;

interface SourceFilePayload {
  filePath: FilePath;
  docs?: Doc[];
}

export interface SerialzieSourceFile {
  filePath: FilePath;
  docs: SerialziedDoc[];
}

export class SourceFile {
  private filePath: FilePath;

  private docs?: Doc[];

  constructor(payload: SourceFilePayload) {
    const { filePath, docs = [] } = payload;
    this.filePath = filePath;
    this.docs = docs;
  }

  public setFilePath(filePath: FilePath): this {
    this.filePath = filePath;
    return this;
  }

  public getFilePath(): FilePath {
    return this.filePath;
  }

  public addDoc(doc: Doc) {
    this.docs!.push(doc);
    return this;
  }

  public static serialize(curr: SourceFile) {
    return {
      filePath: curr.filePath,
      docs: curr.docs?.map(Doc.serialize),
    };
  }
}
