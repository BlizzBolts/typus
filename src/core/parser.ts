import ts from "typescript";
import { generateDoc, generateDocByInternalSymbol } from "../utils/doc";
import { Doc } from "./doc";
import { FilePath, SourceFile } from "./sourceFile";

export const defaultCompilerOptions = {
  jsx: ts.JsxEmit.React,
  module: ts.ModuleKind.CommonJS,
  target: ts.ScriptTarget.Latest,
};

export class Parser {
  private compilerOptions: ts.CompilerOptions = defaultCompilerOptions;
  private filePaths: string[] = [];
  private program!: ts.Program;
  private typeChecker!: ts.TypeChecker;
  private cache: Map<FilePath, SourceFile> = new Map<FilePath, SourceFile>();

  constructor(
    filePaths?: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ) {
    if (filePaths) {
      this.setup(filePaths, compilerOptions);
    }
  }

  static of(
    filePaths?: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ) {
    return new Parser(filePaths, compilerOptions);
  }

  getTypeChecker() {
    return this.typeChecker;
  }

  getProgram() {
    return this.program;
  }

  setup(
    filePaths: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ): this {
    this.compilerOptions = Object.assign(
      defaultCompilerOptions,
      compilerOptions
    );
    this.filePaths = Array.isArray(filePaths) ? filePaths : [filePaths];
    this.program = ts.createProgram(this.filePaths, this.compilerOptions);
    this.typeChecker = this.program.getTypeChecker();

    this.cache.clear();

    return this;
  }

  parse() {
    const validSourceFiles = this.program.getSourceFiles().filter((o) => {
      return !o.isDeclarationFile;
    });

    validSourceFiles.forEach((o: ts.SourceFile) => {
      const sourceFile = new SourceFile({ filePath: o.fileName });
      this.cache.set(o.fileName, sourceFile);
      o.forEachChild((p: ts.Node) => {
        const doc = this.traverse(p, sourceFile);
        if (doc) {
          sourceFile.addDoc(doc);
        }
      });
    });

    return this;
  }

  parseInterfaceDeclaration(
    node: ts.InterfaceDeclaration,
    sourceFile: SourceFile
  ) {
    const type = this.typeChecker.getTypeAtLocation(node.name);
    const doc = generateDoc(node, type, this.typeChecker);
    doc.setRoot(sourceFile);

    if (Array.isArray(node.members)) {
      node.members.forEach((memberNode) => {
        const childDoc = this.traverse(memberNode, sourceFile);
        if (childDoc) {
          childDoc.setRoot(sourceFile);
          doc.children?.push(childDoc);
        }
      });
    }
    return doc;
  }

  parseTypeAliasDeclaration(
    node: ts.TypeAliasDeclaration,
    sourceFile: SourceFile
  ): Doc {
    const type = this.typeChecker.getTypeAtLocation(node.name);
    const doc = generateDocByInternalSymbol(node, type, this.typeChecker);
    doc.setRoot(sourceFile);
    return doc;
  }

  parseMethodSignature(node: ts.MethodSignature, sourceFile: SourceFile): Doc {
    const type = this.typeChecker.getTypeAtLocation(node);
    const doc = generateDoc(node, type, this.typeChecker);
    doc.setRoot(sourceFile);

    if (Array.isArray(node.parameters)) {
      node.parameters.forEach((parameter) => {
        const parameterDoc = this.traverse(parameter, sourceFile);
        if (parameterDoc) {
          parameterDoc.setRoot(sourceFile);
          doc.parameters?.push(parameterDoc);
        }
      });
    }
    return doc;
  }

  parseParameter = (
    node: ts.ParameterDeclaration,
    sourceFile: SourceFile
  ): Doc => {
    const type = this.typeChecker.getTypeAtLocation(node);
    const doc = generateDoc(node.name, type, this.typeChecker);
    doc.setRoot(sourceFile);
    return doc;
  };

  parsePropertySignature = (
    node: ts.PropertySignature,
    sourceFile: SourceFile
  ): Doc => {
    const type = this.typeChecker.getTypeAtLocation(node.name!);
    const doc = generateDoc(node.name, type, this.typeChecker);
    doc.setRoot(sourceFile);

    if (node.type && ts.isFunctionTypeNode(node.type)) {
      node.type.parameters.forEach((parameter) => {
        const parameterDoc = this.traverse(parameter, sourceFile);
        if (parameterDoc) {
          parameterDoc.setRoot(sourceFile);
          doc.parameters?.push(parameterDoc);
        }
      });
    }

    // if (node.type && ts.isTypeReferenceNode(node.type)) {
    //   node.type.parameters.forEach((parameter) =>
    //     this.traverse(parameter, doc)
    //   );
    // }
    return doc;
  };

  // parseFunction(node: ts.FunctionDeclaration, parent: Doc) {
  //   const fnType = this.typeChecker.getTypeAtLocation(node);
  //   const Doc = generateDoc(node, fnType.getSymbol()!, this.typeChecker);
  //   parent.children?.push(Doc);
  //   this.cache.set(fnType, Doc);

  //   if (Array.isArray(node.parameters)) {
  //     node.parameters.forEach((parameter) => this.traverse(parameter, Doc));
  //   }
  // }

  // parseInterfaceMember = (node: ts.TypeElement, parent: Doc) => {
  //   const memberType = this.typeChecker.getTypeAtLocation(node!.name!);

  //   if (this.cache.has(memberType)) {
  //     parent.members?.push(this.cache.get(memberType)!);
  //   } else {
  //     const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
  //     parent.members?.push(generateDoc(node, symbol!, this.typeChecker));
  //   }
  // };

  traverse(node: ts.Node, sourceFile: SourceFile): Doc | undefined {
    // if (this.cache.has(this.typeChecker.getTypeAtLocation(node))) {
    // }

    if (ts.isInterfaceDeclaration(node)) {
      console.log("isInterfaceDeclaration!");
      return this.parseInterfaceDeclaration(node, sourceFile);
    }

    if (ts.isMethodSignature(node)) {
      console.log("isMethodSignature!");
      return this.parseMethodSignature(node, sourceFile);
    }

    if (ts.isPropertySignature(node)) {
      console.log("isPropertySignature!");
      return this.parsePropertySignature(node, sourceFile);
    }

    if (ts.isParameter(node)) {
      console.log("isParameter!");
      return this.parseParameter(node, sourceFile);
    }

    // if (ts.isVariableDeclaration(node)) {
    //   console.log("isVariableDeclaration!");
    //   this.parseIndexSignatureDeclaration(node, parent)
    //   return;
    // }

    // if (ts.isFunctionDeclaration(node)) {
    //   console.log("isFunctionDeclaration!");
    //   this.parseFunction(node, parent);
    //   return;
    // }

    // if (ts.isClassDeclaration(node)) {
    //   console.log("isClassDeclaration");
    //   return;
    // }

    if (ts.isTypeAliasDeclaration(node)) {
      console.log("isTypeAliasDeclaration");
      return this.parseTypeAliasDeclaration(node, sourceFile);
    }

    // if (ts.isEnumDeclaration(node)) {
    //   console.log("isEnumDeclaration");
    //   return;
    // }
    // if (ts.isModuleDeclaration(node)) {
    //   console.log("isModuleDeclaration");
    //   return;
    // }

    // if (ts.isTypeElement(node)) {
    //   console.log("isTypeElement!");
    //   this.parseInterfaceMember(node, parent);
    //   return;
    // }
  }

  serialize() {
    const cacheObject = Object.fromEntries(this.cache.entries());
    return Object.keys(cacheObject).reduce((memo, key) => {
      return {
        ...memo,
        [key]: SourceFile.serialize(cacheObject[key]),
      };
    }, {});
  }
}
