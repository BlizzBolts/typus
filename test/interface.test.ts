import { suite, it, expect } from "vitest";
import path from "path";
import { Parser } from "../src/core/parser";

suite("interface", () => {
  it("should parse type complex interface", () => {
    const docCache = Parser.of()
      .setup(path.resolve(__dirname, "./manifest/complex-interface.ts"))
      .parse()
      .serialize();
    expect(docCache).toMatchSnapshot();
  });
});
