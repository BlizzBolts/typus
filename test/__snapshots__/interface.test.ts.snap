// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`interface > should parse type complex interface 1`] = `
{
  "/Users/ming/projj/github.com/BlizzBolts/typus/test/manifest/complex-interface.ts": {
    "docs": [
      {
        "children": [
          {
            "children": [],
            "documentation": "",
            "members": [],
            "name": "content",
            "parameters": [],
            "required": false,
            "tags": [],
            "type": "string",
          },
        ],
        "documentation": "",
        "members": [],
        "name": "LogPayload",
        "parameters": [],
        "required": false,
        "tags": [],
        "type": "any",
      },
      {
        "children": [
          {
            "children": [],
            "documentation": "",
            "members": [],
            "name": "log",
            "parameters": [
              {
                "children": [],
                "documentation": "",
                "members": [],
                "name": "payload",
                "parameters": [],
                "required": false,
                "tags": [],
                "type": "LogPayload",
              },
            ],
            "required": false,
            "tags": [],
            "type": "(payload: LogPayload) => void",
          },
        ],
        "documentation": "",
        "members": [],
        "name": "ActonMeta",
        "parameters": [],
        "required": false,
        "tags": [],
        "type": "any",
      },
      {
        "children": [
          {
            "children": [],
            "documentation": "",
            "members": [],
            "name": "meta",
            "parameters": [],
            "required": false,
            "tags": [],
            "type": "ActonMeta",
          },
        ],
        "documentation": "",
        "members": [],
        "name": "Action",
        "parameters": [],
        "required": false,
        "tags": [],
        "type": "any",
      },
      {
        "children": [
          {
            "children": [],
            "documentation": "",
            "members": [],
            "name": "Action",
            "parameters": [],
            "required": false,
            "tags": [],
            "type": "Action",
          },
        ],
        "documentation": "",
        "members": [],
        "name": "ComplexInterface",
        "parameters": [],
        "required": false,
        "tags": [],
        "type": "any",
      },
    ],
    "filePath": "/Users/ming/projj/github.com/BlizzBolts/typus/test/manifest/complex-interface.ts",
  },
}
`;
